Cloud ObjectStore

VJalili opened the pull request:
The `Cloud` ObjectStore is a provider-agnostic cloud-based storage. In other words, the `Cloud` ObjectStore can read/write from various cloud-based storages without a provider-specific implementation; it leverages on [`CloudBridge`](https://github.com/gvlproject/cloudbridge) which provides a unified access to various providers. 

This PR sets a base for the User-based ObjectStore we have been working on. Initially, the `Cloud` ObjectStore was part of the User-based ObjectStore (PR [#4314](https://github.com/galaxyproject/galaxy/pull/4314)), and it is being submitted as a separate PR following the comments on the PR [#4314](https://github.com/galaxyproject/galaxy/pull/4314).

To simplify the comparison with `S3ObjectStore`, the `Cloud` ObjectStore provides the very same functionality as `S3ObjectStore`, i.e., read/write to AWS S3 buckets. The only difference is: `S3ObjectStore` leverages on `boto` while `Cloud` uses `CloudBridge`. 

pinging @bgruening @nsoranzo and @erasche

nuwang committed:
Integrated File Browsing and Workflow support for GenomeSpace

nuwang committed:
Moved back some code to wrapper and added back data type sniffing.

nuwang committed:
Added support for pushing multiple files from GenomeSpace to Galaxy.

jmchilton committed:
Bring Galaxy, reports, and shed run scripts in sync.

- They now uniformly handle server option processing.
- Reports and Tool Shed now respect many new and important options such as --skip_venv.
- Reports and Tool Shed now do extra checking around Python version that Galaxy does.

In addition to improving the reports and tool shed startup scripts and reducing cognative load with respect to option handling, this is an important pre-condition to us being able us to switch on uwsgi by default for all three services simultaneously without a bunch of copy and pasting.

jmchilton committed:
Centralize launching server for reports and tool shed.

Initial outline of allowing launch via uwsgi.

jmchilton committed:
Update tool shed to use newer configuration schema.

 - Implement kwalify schema to validate/define the schema.
 - Replace ini configuration with YAML (completely backward compatible).
 - Rebuild a sample configuration from the schema (config/tool_shed.yml.sample).

jmchilton committed:
Update galaxy.util.properties to allow YAML.

jmchilton committed:
Scripts to manage configuration infrastructure.

This has been rebased several times and includes a variety of fixes (some thanks to martenson).

jmchilton committed:
Makefile shortcuts for configuration script.

jmchilton committed:
Add uwsgi_schema.yml.

jmchilton committed:
Update Galaxy Reports to use newer configuration schema.

 - Implement kwalify schema to validate/define the schema.
 - Replace ini configuration with YAML (completely backward compatible).
 - Rebuild a sample configuration from the schema (config/reports.yml.sample).

jmchilton committed:
Add reports documentation to ``doc/``.

- Automatically generate RST docs from config schema.
- Describe older functionality as well as new YAML-based configuration.
- Replace https://wiki.galaxyproject.org/Admin/UsageReports.

jmchilton committed:
Add required dependency on uwsgi.

Required for the new out-of-the-box configuration of tool shed and reports.

jmchilton committed:
Add pykwalify as dependency for the new config validation stuff.

xref https://github.com/galaxyproject/starforge/pull/110

jmchilton committed:
Fix Python 3 linting problem with test_config_manage.py.

Slugger70 committed:
Fixed genomespace_importer.xml command block

jmchilton committed:
Pin version of uWSGI to freshly released 2.0.15 available on wheels.galaxyproject.org.

nuwang committed:
Fixed pylint errors

jmchilton committed:
Fix ORM scripts for #3179.

jmchilton committed:
Update config overhaul branch for introduction of galaxy.web.stack package.

nsoranzo committed:
Use container-based infrastructure (sudo: false) in TravisCI

bernt-matthias committed:
added missing parameter to function ensure_installed

bernt-matthias committed:
ensure_installed is now iterative instead of recursive

bernt-matthias committed:
version with no repetition at all

bernt-matthias committed:
Set timeout for aquiring lock to 5min

jmchilton committed:
Improve error message for unknown module types when encountered.

jmchilton committed:
Allow API tests to work with admin users (not just admin keys).

jmchilton committed:
Print warning if any input connections are unused during workflow execution.

VJalili committed:
Rebase.

VJalili committed:
Added CloudBridge to the requirements list.

VJalili committed:
Added CloudBridge to the dependencies check.

VJalili committed:
(1) Added Cloud to ObjectStore; (2) updated CloudBridge dependencies.

VJalili committed:
- Updated `Cloud` to account for changes in `CloudBridge` interface.
- Added `Cloud` to objectstore import.

VJalili committed:
Reverted the changes on package versions, as these changes are
requests via a different PR.

mvdbeek commented on lib/galaxy/objectstore/cloud.py:
```diff
@@ -0,0 +1,615 @@
+    import boto
+
+    from boto.exception import S3ResponseError
+    from boto.s3.key import Key
```
The linter doesn't like unused imports (this line and the following), but you can ignore this by doing
```
from boto.s3.key import Key  # noqa: F401
from boto.s3.connection import S3Connection  # noqa: F401
```

Although I'm not sure in which scenario it would help to import unused dependencies ...
If external code needs it, it should probably import the dependencies where they are needed.

VJalili replied:
@mvdbeek Thanks, the imports are updated accordingly. 
Actually, `boto` import is temporary solution for handling AWS S3 exceptions until `cloudbridge` implements such exceptions for all the providers.

VJalili committed:
Changed boto import style.

nuwang committed:
Fix incorrect reference when determining file type through file ext

afgane commented on lib/galaxy/dependencies/conditional-requirements.txt:
```diff
@@ -13,6 +13,7 @@ graphitesend
 azure-storage==0.32.0
 # PyRods not in PyPI
 python-ldap==2.4.27
+cloudbridge==0.3.1
```
Any reason not to go with the latest current release (0.3.2)?

VJalili replied:
No specific reason, I updated to the current latest version (also see starforge PR [#138](https://github.com/galaxyproject/starforge/pull/138)).

afgane commented on lib/galaxy/objectstore/cloud.py:
```diff
@@ -0,0 +1,611 @@
+from ..objectstore import convert_bytes, ObjectStore
+from cloudbridge.cloud.factory import CloudProviderFactory, ProviderList
+
+try:
```
Presumably, this is for backward compatibility? Either way, it'd be good to have a comment why is this necessary in addition to the CloudBridge import.

VJalili replied:
`boto` was only used to handle some exceptions. However, I replaced all `boto`-specific exceptions with a generic exception. This is totally a temporary patch till exceptions are properly wrapped and thrown by `CloudBridge`.

afgane commented on lib/galaxy/objectstore/cloud.py:
```diff
@@ -0,0 +1,611 @@
+                    bucket = self.conn.object_store.create(bucket_name)
+                log.debug("Using cloud object store with bucket '%s'", bucket.name)
+                return bucket
+            except S3ResponseError:
```
This is not imported anywhere. Also, `create` method returns `S3CreateError` (http://boto.cloudhackers.com/en/latest/ref/boto.html#module-boto.exception). 
More importantly, this will not work cross-provider so having an extra, general `except` may be desirable?
Same comment applies to other refs to `S3ResponseError`.
(xref https://github.com/gvlproject/cloudbridge/issues/9 and custom/standardized CloudBridge exceptions).

VJalili replied:
I agree, and as explained in the previous comments, as a temporary patch, replaced `S3`/`boto`-specific exceptions with generic exceptions.

afgane commented on lib/galaxy/objectstore/cloud.py:
```diff
@@ -0,0 +1,611 @@
+            self.secret_key = a_xml.get('secret_key')
+            b_xml = config_xml.findall('bucket')[0]
+            self.bucket = b_xml.get('name')
+            self.use_rr = string_as_bool(b_xml.get('use_reduced_redundancy', "False"))
```
It doesn't look like we're using this option anymore (which is understandable from the multi-cloud perspective), so probably best to remove this line and relevant config.

afgane commented on lib/galaxy/objectstore/cloud.py:
```diff
@@ -0,0 +1,611 @@
+                        created_obj = self.bucket.create_object(rel_path)
+                        created_obj.upload(source_file)
+                    else:
+                        self.bucket.get(rel_path).upload(source_file)
```
Should probably keep the same code structure as in the `if` case (or the other way around, as long as it's consistent).

afgane commented on lib/galaxy/objectstore/cloud.py:
```diff
@@ -0,0 +1,611 @@
+                              rel_path)
+                    return True
+                # FIXME: don't need to differenciate between uploading from a string or file,
+                # because CloudBridge handles this internally.
```
What's there to fix? `if` and `else` here look the same so shouldn't it just be consolidated?

VJalili replied:
Updated. But I'm postponing function signature change for future, just to keep the changes at this PR at a minimal possible level.

afgane commented on lib/galaxy/objectstore/cloud.py:
```diff
@@ -0,0 +1,611 @@
+            if not os.path.exists(cache_dir):
+                os.makedirs(cache_dir)
+
+            # Although not really necessary to create S3 folders (because S3 has
```
Should probably remove this comment.

afgane commented on lib/galaxy/objectstore/cloud.py:
```diff
@@ -0,0 +1,611 @@
+            if entire_dir and extra_dir:
+                shutil.rmtree(self._get_cache_path(rel_path))
+                results = self.bucket.list(prefix=rel_path)
+                for key in results:
```
May be better to replace references like this that use AWS terminology to something cloud agnostic? e.g., `object`. This applies throughout the file (e.g., `_key_exists`).

VJalili replied:
Agree. However, as I mentioned in other comments, I'm trying to keep `Cloud` very similar to `S3ObjectStore` (following the comments on the PR #4314). Using a general terminology might escalate the challenges :) But we'll surely use more general nomenclature in future updates.

afgane reviewed (changes requested):
Looks pretty good. Clearly, this is still AWS-only even if it implements the CloudBridge interface but I guess that's just the first step. 
There are a few comment lines and debug print statements that were also in the previous implementation. I think those can be removed now. Finally, I made a few comments inline, some of which repeat themselves in multiple places but the comment should apply to whole the file.

VJalili committed:
Updated CloudBridge to its current latest version.

VJalili committed:
Commented on boto import, and removed a `use_reduced_redundancy` config.

VJalili committed:
Removed boto import, and replaced all S3-specific exception catches with
a generic exception--a temporary solution till CloudBridge wraps the
exceptions properly.

VJalili committed:
Consolidated two upload methods, i.e., upload from file and string,
because CloudBridge internally maps to appropriate functions depending
on the input type.

VJalili committed:
Removed a S3-related comment, and swapped an if/else condition.

VJalili committed:
Removed a comment section.

VJalili commented:
@afgane Thanks for the comments, I updated the code accordingly. Few points: 

> Clearly, this is still AWS-only even if it implements the CloudBridge interface but I guess that's just the first step.

Correct. This PR sets the ground for User-based ObjectStore (initially discussed at #4314); which is a user-based and provider-agnostic ObjectStore. However, following the comments at #4314, I made `Cloud` with least possible changes to `S3ObjectStore` to (a) simplify comparison, (b) incremental and smaller impacts on the Galaxy codebase. Accordingly, you may see lots of `S3`-specific terminology and comments. I think we could be more provider-agnostic in future updates. 

I'll comment inline other comments.

afgane commented:
It's generally not great to use the generic `Exception` if it can be helped but given this is an intermediary step, I think it's ok.

There are a couple of tests still failing though. How do the dependencies that are still getting built (i.e., the open PR for CloudBridge wheel) get pulled into the test env? Could that be it?

VJalili commented:
Maybe the tests are failing because the current latest version of the `CloudBridge` is not built yet (https://github.com/galaxyproject/starforge/pull/138)

mvdbeek committed:
Show verison button in workflow editor and trigger on_change

mvdbeek committed:
Do postchange action only when modifiying tool versions in the workflow editor

mvdbeek committed:
Fix qunit test

mvdbeek committed:
Keep spacing consistent (thx @nsoranzo)

mvdbeek committed:
Reset outputs when changing tool versions in wf-editor

mvdbeek committed:
Now thing are mostly working -- input and output connections are correctly altered ... but saving the workflow and reloading occasionally loads the wrong workflow

mvdbeek committed:
Fix the occasional loading of the wrong workflow. Turns out that sometimes the tool_version is undefined (i.e) if the exact node had not been manipulated, but config_form.version is always up to date

VJalili committed:
Changed indentation of two lines.

mvdbeek committed:
Now adding/removing datasets and multiple version switches work fine, but PJA parameters are still problematic

mvdbeek committed:
Fix logic for adding and removing outputs

mvdbeek committed:
Remove extra newlines

mvdbeek committed:
Remove unused old_version variable

mvdbeek committed:
Fix typo and remove leftover comment (thanks @nsoranzo)

nsoranzo commented:
[Unit tests](https://travis-ci.org/galaxyproject/galaxy/jobs/261703367) are failing with:
```
File "/home/travis/build/galaxyproject/galaxy/lib/galaxy/objectstore/cloud.py", line 25, in <module>

    from cloudbridge.cloud.factory import CloudProviderFactory, ProviderList

ImportError: No module named cloudbridge.cloud.factory
```

mvdbeek committed:
Remove unused workflow outputs

hexylena committed:
Also load yaml sample files. cc @jmchilton, is this ok?

hexylena committed:
correct name to type

nuwang committed:
Improved recognition of compressed files

jmchilton committed:
Cleanup a couple things in #4396 caught by @nsoranzo.

jmchilton committed:
Generalize allow_library_path_paste to allow_path_paste.

Enable pasting paths in the history upload GUI & general purpose tools API for admin users if enabled.

jmchilton committed:
Add configuration option to disable upload content checking.

jmchilton committed:
Better error reporting in test/base/populators.py.

jmchilton committed:
Add ability to arbitrarily not decompress files on upload via API.

jmchilton committed:
Configuration tests for the last few upload enhancements.

hexylena committed:
fix import order

jmchilton committed:
Remove unneeded messages.mako.

nsoranzo committed:
Update conda channel order to sync with Bioconda

hexylena committed:
fix css class

hexylena committed:
not a useful test in this location

hexylena committed:
rebuild client

jmchilton committed:
Merge pull request #4305 from erasche/bug-report-actions

Refactor bug reporting into plugins / Remove bug report mako

jmchilton committed:
Merge pull request #4408 from erasche/grt-pardir

Small GRT bugfixes

jmchilton committed:
Remove unused dataset_info in parameters.grouping.

jmchilton committed:
Merge pull request #4329 from glormph/collection_list_sort

Created alphabetic and numerical sort tool for collection operations

jmchilton committed:
Merge pull request #4254 from ffinfo/vcf_gz_datatype

Vcf gz datatype

guerler committed:
Merge branch 'dev' into revise_grid_forms

guerler committed:
Merge branch 'revise_grid_forms' into admin_grid_roles

guerler committed:
Merge branch 'admin_grid_roles' into admin_grid_groups

guerler committed:
Fix lint

jmchilton committed:
Merge remote-tracking branch 'jmchilton/dev' into install_workflow_and_tools

Conflicts:
 lib/galaxy/webapps/galaxy/api/workflows.py

jmchilton committed:
Require admin for new import_tools function in workflows API.

Includes test case.

jmchilton committed:
Merge pull request #4039 from VJalili/patch-11

Updated parameters check for OpenID

guerler committed:
Merge branch 'admin_grid_groups' into admin_grid_toolversion

nsoranzo committed:
Add missing chemical formats to datatypes_conf.xml.sample

nsoranzo committed:
Update pin for flake8-docstrings

Similar to commit 2586df45637d794c63252838573b2d1e6d692243 .

Commit:
Remove spurious null check.

Commit:
Add missing `var` declarations.

In all these cases, a variable was not declared to be a local using `var` (and hence was implicitly interpreted as a global variable), but all uses of that variable were preceded by a definition within the same function, suggesting that the variable was, in fact, meant to be local, not global.

guerler committed:
Fix lint

martenson committed:
Merge pull request #4377 from guerler/revise_grid_forms

Remove makos of admin user grid forms

jmchilton committed:
Merge pull request #4385 from guerler/admin_grid_roles

Admin grid roles revision

jmchilton committed:
Tweak Excel format phrasing based on review from @nsoranzo.

jmchilton committed:
Add EDAM format for excel.xls datatype.

martenson committed:
Merge pull request #4404 from jmchilton/allow_path_paste

Generalize allow_library_path_paste to allow_path_paste.

bgruening committed:
Merge pull request #4410 from jmchilton/excel97

Add Excel97 Datatype

jmchilton committed:
Fix bug in upload HTML checker.

If a chunk was passed into the check_html function - e.g. it was read from a compressed file - it would search through each character in the string for certain HTML content instead of each line in the string.

bgruening committed:
Merge pull request #3179 from jmchilton/web_config_overhaul_0

Configuration & Server Overhaul - Tool Shed and Reports

jmchilton committed:
Merge pull request #3064 from manabuishii/install_workflow_and_tools

Install tools and workflow

mvdbeek committed:
Don't cast tool_version to string if tool_version is None-type

If a version is not specifically requested a wrong version (`None` cast to
`'None'`) of a tool will be loaded in the workflow editor instead of the
latest version.

That should address @martenson's remark in
https://github.com/galaxyproject/galaxy/pull/4373#issuecomment-321896674
(and maybe also
https://github.com/galaxyproject/galaxy/issues/557#issuecomment-213498153).

mvdbeek committed:
Add static JS

Commit:
Remove an unused variable.

The result of the `_.map` call was never used, so I replaced it with `_.each`.

bgruening committed:
Merge pull request #4417 from jmchilton/upload_enhance

Fixes and enhancements for the upload API

mvdbeek committed:
Rebuild tool_version_cache on DetachedInstanceError

mvdbeek committed:
Merge pull request #4397 from jmchilton/api_test_enhancement

Better Error Summaries for API Tests

mvdbeek committed:
Merge pull request #4394 from jmchilton/subworkflow_loading_fix

Bugfix for loading subworkflows from workflow descriptions.

mvdbeek committed:
Merge pull request #4413 from nsoranzo/release_17.01_really_add_datatypes

[17.01] Add missing chemical formats to datatypes_conf.xml.sample

nsoranzo committed:
Remove _ensure_conda_context() function

and make conda_context argument mandatory in functions/methods.
This should help making sure the correct context (e.g. channels)
is used for all Conda operations.

Also remove unused is_target_available() function.

nsoranzo committed:
Merge branch 'release_17.01' into release_17.05

Conflicts:
	tox.ini

nsoranzo committed:
Merge branch 'release_17.05' into dev

Conflicts:
	static/maps/mvc/library/library-foldertoolbar-view.js.map
	static/scripts/bundled/analysis.bundled.js.map
	static/scripts/bundled/libs.bundled.js.map
	static/scripts/mvc/library/library-foldertoolbar-view.js

nsoranzo committed:
Override channel list on the command line instead of modifying condarc

The way we were modifying condarc in the ensure_channels_configured()
method was broken because we were appending the missing channels at the
end, so the final order may not be the one specified in
conda_ensure_channels .

Also:
- not modifying the user condarc is a much better approach
- overriding the channel list on the command line will make debugging
  easier

guerler committed:
Removed unreached message line

guerler committed:
Merge branch 'dev' into admin_grid_groups

nsoranzo committed:
Use log.warning() instead of deprecated warn()

nsoranzo committed:
Remove container config log message from cleanup scripts logs

hexylena committed:
some unicoding

jmchilton committed:
Merge pull request #4420 from mvdbeek/fix_wrong_tool_version_cast

[17.05] Don't cast tool_version to string if tool_version is None-type

jmchilton committed:
Merge pull request #4426 from erasche/local-runner-unicode

some unicoding for local runner

jmchilton committed:
Merge pull request #4409 from nsoranzo/fix_conda_channel_order

Update conda channel order to sync with Bioconda

nsoranzo committed:
Merge branch 'release_17.05' into dev

Conflicts:
	lib/galaxy/model/tool_shed_install/__init__.py

martenson committed:
Merge branch 'dev' into switch_versions_in_workflows

martenson committed:
Merge pull request #4373 from mvdbeek/switch_versions_in_workflows

Allow switching tool versions in workflow editor

dannon committed:
Merge pull request #4427 from nsoranzo/merge_release_17.05

Merge release_17.05 into dev

guerler committed:
Merge branch 'dev' into admin_grid_groups

guerler committed:
Merge branch 'admin_grid_groups' into admin_grid_toolversion

guerler committed:
Move form augmentation to form handler

nsoranzo committed:
Fix import order and Python3 compatibility

guerler committed:
Fix

Fix

gregvonkuster committed:
Add support for the MEME psp format

martenson committed:
Merge pull request #4405 from guerler/admin_grid_groups

Admin grid groups revision

guerler committed:
Optionally allow to show or hide select all

guerler committed:
Hide select all for role and group selectors

guerler committed:
Use 'Save' as default title for configuration forms

guerler committed:
Fix user id decoding, fix optional array handling

guerler committed:
Fix label for default quotas

mvdbeek committed:
Break connection in workflow editor if necessary

When switching tool versions in the workflow editor connections are
maintained by default. If a tool changes the datatype one may end up
with a connection between incompatible input/outputs. Therefore we
update the datatypes and explicitly destroy invalid connections.

guerler committed:
Fix messages

guerler committed:
Catch user input error for auto created roles and groups

guerler committed:
Move decode id to correct location

guerler committed:
More messages to validate inputs

guerler committed:
Add validation messages if resource ids are not correct

guerler committed:
Fix private role validation

guerler committed:
Merge branch 'dev' into admin_grid_toolversion

jmchilton committed:
Merge pull request #4431 from mvdbeek/break_connections

Break connection in workflow editor if necessary

bgruening committed:
fix XML error

guerler committed:
Separate input augmentation helpers from form views, augment inputs for regular form

jmchilton committed:
Overhaul of tool provided job metadata.

- Add another test to demonstrate referencing entries in galaxy.json for datasets by dataset path basename instead of ID.
- Add a bit more documentation to test tools.
- Redo the structure of galaxy.json for profile >= 17.09 tools with two new test tools to demonstrate the new structure - for both output dataset metadata and discovered dataset metadata.
- Improved abstractions for interaction between tools and jobs to enable different tool provided metadata collection depending on tool profile version.

bgruening committed:
Merge pull request #4387 from jmchilton/tool_unit_test_cleanup

Cleanup of Unit Test Code for Loading Tools

guerler committed:
Simplify form wrapper calling

dannon committed:
Merge pull request #4432 from guerler/fix_role_association_check

Fix private role validation

gregvonkuster committed:
Change Psp to MemePsp and misc fixes per initial review

Commit:
Move two variable definitions to combine them with their declarations.

Commit:
Run `make client`.

gregvonkuster committed:
Remove strange import - no idea from where it originated

dannon committed:
Merge pull request #4421 from xiemaisi/fix-lgtm-alerts

Fix lgtm alerts

chambm committed:
Avoid adding `(imported from API)` to workflows

This should fail a few tests which I will fix with additional commits. Doing this from GitHub...

chambm committed:
Remove `(imported from API)` suffix from test

chambm committed:
Remove `(imported from API)` suffix from test

jmchilton committed:
Allow discovering datasets directly from galaxy.json...

without needing to specify a discovered dataset pattern. If the ``discovered_datasets`` tag includes ``from_tool_provided_metadata="true"`` the datsets listed in galaxy.json will just be used directly without needing to be "discovered" using a dataset pattern. Metadata and such can continue to be specified either in that file or on the discovered_dataset element except things like pattern (which is not used) and sort_by (since the json should describe the order I suppose).

jmchilton committed:
Allow discovering dataset collections in galaxy.json in a nested fashion.

Fixes #2658.

guerler committed:
Restore build model bypass

Pack

guerler committed:
Merge branch 'dev' into improve_form_modules

guerler committed:
Attempt to customize build model

guerler committed:
Remove unfilterable columns from group and quota search grids

martenson committed:
Merge branch 'dev' into admin_grid_toolversion
and repack

martenson committed:
Merge pull request #4411 from guerler/admin_grid_toolversion

replace mako templates for tool lineage and quotas grids

chambm committed:
Remove print of non-existent attribute 'content'

guerler committed:
Decouple model building

gregvonkuster committed:
Fix MemePsp sniff entry

guerler committed:
Hide initial tool version message

guerler committed:
Move workflow tool form model rebuilding to workflow forms

guerler committed:
Properly update workflow tool module

guerler committed:
Add optional initial model caller

Pack

guerler committed:
Remove form dom removal through workflow manager

guerler committed:
Merge branch 'dev' into improve_form_modules

guerler committed:
Fix message style

nsoranzo committed:
Fix all E201 and E202 style errors

using the following command:
```
autopep8 -i -r --exclude $(sed -e 's|^|./|' -e 's|/$||' .ci/flake8_blacklist.txt | paste -sd,) --select E201,E202 .
```

nsoranzo committed:
Fix E127 errors introduced by previous commit

using:
```
autopep8 -i -r --exclude $(sed -e 's|^|./|' -e 's|/$||' .ci/flake8_blacklist.txt | paste -sd,) --select E127 .
```

jmchilton committed:
Move determining tool provided metadata style to tool parser.

Allow override by specifying "provided_metadata_style=" attribute on outputs in tool XML - this should allow newer profile tools to use the older style JSON providing a clear path for upgrades.

nsoranzo committed:
Fix more indentation errors

jmchilton committed:
Allow override of provided metadata location.

Eliminates that terrible circular dependency between tools and jobs.

bgruening committed:
Merge pull request #4439 from chambm/patch-2

Remove print of non-existent attribute 'content'

bgruening committed:
Merge pull request #4434 from chambm/patch-1

Avoid adding `(imported from API)` to workflows

bgruening committed:
change file_ext to be inline with memexml and co

bgruening commented:
@galaxybot test this

VJalili commented:
@bgruening the test will fail because of a missing wheel (see [this PR](https://github.com/galaxyproject/starforge/pull/139)).

VJalili committed:
Update CloudBridge to its current latest

Should use the wheel created at [this PR](https://github.com/galaxyproject/starforge/pull/139).

VJalili committed:
Update CloudBridge to its current latest

Should use the wheel created at [this PR](https://github.com/galaxyproject/starforge/pull/139).

jmchilton committed:
Variety of fixes for #4437.

jmchilton committed:
Merge pull request #4440 from nsoranzo/fix_E201_E202

Fix E201 and E202 style errors

nsoranzo committed:
Fix API tests broken in #4434

guerler committed:
Fix help text style

dannon committed:
Merge pull request #4445 from nsoranzo/followup_on_4434

Fix API tests broken in #4434

dannon committed:
Merge pull request #4430 from gregvonkuster/psp_format

Add support for the MEME psp format

dannon committed:
Fix E201 errors (spaces in parens) in new memepsp datatype

nsoranzo committed:
Merge pull request #4446 from dannon/memepsp_formatting

Fix E201 errors (spaces in parens) in new memepsp datatype

nuwang committed:
Added new config setting for GenomeSpace UI URL

guerler committed:
Fix title

nuwang committed:
Fixed typo in exporter help text

guerler committed:
Preserve advanced search visibility

jmchilton committed:
Add autopep8 script to help rebasing branches after #4440.

Fixes a few more PEP8 issues as part of the autopep8'ing as well.

guerler committed:
Fix filters

guerler committed:
Prevent search tag duplication

dannon committed:
Merge pull request #4450 from mvdbeek/fix_file_lock_error

[17.01] Fix file lock error handling

mvdbeek committed:
Merge remote-tracking branch 'origin/release_17.01' into release_17.05

nsoranzo committed:
Merge branch 'release_17.05' into dev

dannon committed:
Merge pull request #4447 from jmchilton/autopep8

Add autopep8 script to help rebasing branches after #4440.

martenson committed:
Merge pull request #4449 from guerler/fix_grid_search

Fixes for advanced grid search

nsoranzo committed:
Uniform indentation of multiline if conditionals

Use 8-spaces indentation for the conditional continuation lines of an
`if`-statement where the conditional part spans multiple lines.

According to PEP-8, using 4-spaces "can produce a visual conflict with
the indented suite of code nested inside the `if`-statement", but the PEP
"takes no explicit position".

There were only 2 places in the whole codebase where 8 spaces were not
used, this will make it uniform.

guerler committed:
Merge branch 'dev' into improve_form_modules

nuwang committed:
Formatting and linting fixes for GenomeSpace tool

VJalili committed:
Add Cloud configuration to object_store_conf.xml.sample

VJalili committed:
Add Cloud configuration to object_store_conf.xml.sample

xgaia committed:
possibility to don't uppercase env variables

nuwang committed:
Fixed alignment of buttons in Genomespace Importer

nsoranzo committed:
Fix import order and Python3 compatibility for lib/galaxy/util/pastescript/serve.py

nsoranzo committed:
Merge pull request #4396 from jmchilton/workflow_warnings

Slightly improve workflow warnings/errors logging.

hexylena committed:
Merge pull request #4454 from xgaia/interactive_environment

Interactive Environment: possibility to don't uppercase env variables

hexylena committed:
More detailed sentry reports

mvdbeek committed:
Merge pull request #4437 from jmchilton/galaxy_json_rev

Enhanced tool options for dataset discovery

hexylena committed:
swap id to username

xgaia committed:
don't uppercase envoverride

nuwang committed:
Fixed template formatting according to review comments

jmchilton committed:
Merge pull request #4456 from nsoranzo/logging_fixes

Logging fixes

dannon committed:
Merge pull request #4455 from nsoranzo/multiline_if_conditionals

Uniform indentation of multiline if conditionals

nuwang committed:
Moved genomespace token to environment variable

jmchilton committed:
Merge pull request #4457 from erasche/sentry-extra-info

More detailed sentry reports

nuwang committed:
Registered genomespace_importer as a tool needing Galaxy's python env

mvdbeek committed:
Fix checking WorkflowInvocation for published workflows

dannon committed:
Merge pull request #4438 from guerler/improve_form_modules

Augment form module separation

dannon committed:
Client rebuild, @guerler and I were trying to figure out what changed in the recently merged #4438 but don't see anything obvious.

martenson committed:
Merge pull request #4464 from dannon/rebuild_000

Client rebuild, @guerler and I were trying to figure out what changed…

nuwang committed:
Rebuilt galaxy client files

martenson committed:
Merge pull request #1814 from gvlproject/genomespace_workflows

Galaxy Workflow support for GenomeSpace

mvdbeek committed:
Correctly raise Exception if invocation not found

mvdbeek committed:
Fix test to use invocation_id instead of history_id

nuwang commented on lib/galaxy/objectstore/cloud.py:
```diff
@@ -0,0 +1,565 @@
+    def _get_bucket(self, bucket_name):
+        """ Sometimes a handle to a bucket is not established right away so try
+        it a few times. Raise error if connection is not established. """
+        for i in range(5):
```
@VJalili Can you elaborate on why this is needed and when it's triggered? Is there some parallel creation of a bucket going on? If not, I think this should be handled at the cloudbridge level, what do you think @afgane?

nuwang commented on lib/galaxy/objectstore/cloud.py:
```diff
@@ -0,0 +1,565 @@
+                log.debug("Parallel pulled key '%s' into cache to %s", rel_path, self._get_cache_path(rel_path))
+                ncores = multiprocessing.cpu_count()
+                url = key.generate_url(7200)
+                ret_code = subprocess.call("axel -a -n %s '%s'" % (ncores, url))
```
@VJalili @afgane This axel based downloading looks interesting - I wonder whether we should consider integrating similar functionality into cloudbridge? Not necessarily through axel, but at least a better implementation than the single threaded one we have now - maybe using an asyncio loop?

afgane commented:
There is an issue with python 2.7 and asyncio but native download optimization would be nice.

null

mvdbeek committed:
Check ownership of invocation, instead of stored_workflow

mvdbeek committed:
Allow admin user to view workflow invocation (thx @nsoranzo)

mvdbeek committed:
Only skip check_security checks if check_ownership and check_accessible are false

mvdbeek committed:
Fix subworkflow tests

mvdbeek committed:
Show up to 15 histories by default in saved histories menu

mvdbeek committed:
Show tag if < 3 tags on history

martenson committed:
Merge pull request #4467 from mvdbeek/saved_history_tags

More items per page in saved history and directly show tags

jmchilton committed:
Merge pull request #4465 from mvdbeek/workflow_invocation_security_check

Fix checking WorkflowInvocation for published workflows

guerler committed:
Fix quota function name

nsoranzo committed:
Merge pull request #4469 from guerler/fix_quota_api

Fix quota function name

martenson committed:
replace exclamation with a clock icon

dannon committed:
Merge pull request #4485 from martenson/remove-exclamation

replace exclamation with a clock icon

VJalili committed:
Removed white spaces.

VJalili committed:
Removed white spaces.

VJalili committed:
Merge remote-tracking branch 'origin/CloudObjectStore' into CloudObjectStore

# Conflicts:
#	lib/galaxy/objectstore/cloud.py

hexylena committed:
Merge pull request #4458 from xgaia/interactive_environment

don't uppercase envoverride

VJalili committed:
Merge branch 'CloudObjectStore' of https://github.com/VJalili/galaxy into CloudObjectStore

VJalili commented:
Some thing went awfully wrong with `rebase`. Please follow up #4487

VJalili commented:
@nuwang @afgane due to an issue with this branch due to a white-space removal rebase, please follow up on a new PR #4487
